File: js/add-binary.js
// https://leetcode.com/problems/add-binary/
// Related Topics: Math, String
// Difficulty: Easy

/*
Initial thoughts:
Looping over both arrays from the end backwards we are going to
add the corresponding digits of both of the arrays (if there is none for one of them then zero),
all the while taking care of a possible remainder and carrying it to the next left digit.
At the end we need to take care of a potential remainder and add it to the array.

For performance purposes, we are going to create the result in a reversed array and reverse it
at the end.
That is because pushing to an array is way more efficient than unshifting. A single reversal at
the end of the algorithm is just O(n) and doable.
Another workaround would be to use a doubly linked list and convert it to an array at the end
of the algorithm which would also be no more than an additional O(n).

Time complexity: O(Max(n,m)) where n === a.length and m === b.length
Space complexity: O(Max(n,m)) where n === a.length and m === b.length

*/

/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
const addBinary = (a, b) => {
  const result = [];
  let carry = 0;

  for (let i = a.length - 1, j = b.length - 1; i >= 0 || j >= 0; i--, j--) {
    let res = (Number(a[i]) | 0) + (Number(b[j]) | 0) + carry;
    carry = (res / 2) | 0;
    res = res % 2;
    result.push(res);
  }
  if (carry) result.push(carry);
  return result.reverse().join('');
};

File: python/add-binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        res = []
        carry, i, j = 0, len(a)-1, len(b)-1
        
        while i >= 0 or j >= 0:
            curr = carry
            if i >= 0:
                curr += int(a[i])
                i -= 1
            if j >= 0:
                curr += int(b[j])
                j -= 1
            res.append(curr%2)
            carry = curr // 2
        
        if carry:
            res.append(carry)
        return "".join(map(str, reversed(res)))
    
    def addBinary2(self, a: str, b: str) -> str:
        res = []
        carry, i, j = 0, len(a)-1, len(b)-1
        
        while i >= 0 and j >= 0:
            curr = carry + int(a[i]) + int(b[j])
            res.append(curr%2)
            carry = curr // 2
            i, j = i-1, j-1
        
        while i >= 0:
            curr = carry + int(a[i])
            res.append(curr%2)
            carry = curr // 2
            i -= 1
        
        while j >= 0:
            curr = carry + int(b[j])
            res.append(curr%2)
            carry = curr // 2
            j -= 1
        
        if carry:
            res.append(carry)
        return "".join(map(str, reversed(res)))